core/trino-main/src/main/java/io/trino/server/security/oauth2/OAuth2Service.java

```diff
@@ -277,6 +289,25 @@ else if (nonce.isPresent() != oauth2Response.getIdToken().isPresent()) {
     private Optional<Claims> internalConvertTokenToClaims(String accessToken)
             throws ChallengeFailedException
     {
+        if (userinfoUri.isPresent()) {
+            // validate access token is trusted by remote userinfo endpoint
+            Request request = Request.builder()
+                    .setMethod(POST)
+                    .addHeader(AUTHORIZATION, "Bearer " + accessToken)
+                    .setUri(userinfoUri.get())
+                    .build();
+            try {
+                Map<String, Object> userinfoClaims = httpClient.execute(request, JsonResponseHandler.createJsonResponseHandler(JsonCodec.mapJsonCodec(String.class, Object.class)));
```

Comment:
Without any caching?

Reply:
The response from `/userinfo` does not necessarily contain an expiration (e.g. Hydra does not provide one), so there's no consistent way to set the expiration for that information. We could add some caching for this in the future if we think it necessary. Even called on every query the volume of calls to `/userinfo` is relatively low. There's also a small benefit of always retrieving the latest response as the user will be denied as soon as their token expires.